utils/visit.test.js

```diff
@@ -0,0 +1,73 @@
+const assert = require('assert');
+
+const visit = require('./visit');
+const { walk } = require('./walk');
+
+describe('visit()', function () {
+  it('runs the function on all features if no other entry point is specified', function () {
+    const walker = walk('html.elements.a');
+    visit(
+      'html.elements.a',
+      () => true,
+      visitorPath => {
+        assert.strictEqual(visitorPath, walker.next().value.path);
+      },
+    );
+  });
+
+  it('skips features not selected by testFn', function () {
+    const hits = new Set();
+    const misses = new Set();
+    visit(
+      'css',
+      path => {
+        if (path.includes('at-rules')) {
+          return true;
+        }
+        misses.add(path);
+        return false;
+      },
+      path => {
+        hits.add(path);
+      },
+    );
+
+    for (const miss in misses) {
+      assert.ok(!hits.has(miss));
+    }
+  });
+
+  it('visitorFn can break iteration', function () {
+    visit(
+      undefined,
+      () => true,
```

Comment:
This highlights two ugly things about `visit()` right now:

- To walk the whole tree, you pass in `undefined` as the first argument. That's not exactly an obvious notation. Maybe it'd be friendlier if you passed in another symbol such as `visit.ROOT` or a catchall string (e.g., `*`).
- To visit every feature, you pass in a function that returns `true`. That's more obvious than the first argument's `undefined`, but still not super cryptic. Similar solutions are available here such as a `visit.ALL` or special casing truthy, non-function values, as in `visit(undefined, true, visitorFn)`.

Taken together, there might be an argument to be made for handling multiple signatures for `visit()`, though that gets complex fast:

- `visit(visitorFn)`  (starts at the root, no test — roughly equivalent to `walk()`) 
- `visit(testFn, visitorFn)` (starts at the root, visits those features that pass tests)
- `visit(str, testFn, visitorFn)` (explicit start point)
- `visit(str, visitorFn)` (explicit start point, no test)

Reply:
I'm glad I wrote this out and had to add the `data` argument. I had a bit of an insight here: the visitor function is the only part that's required and the rest are options. I've refactored this to a new-and-improved signature:

`visit(visitor, options)`

or, fully expanded: `visit(visitor, { entryPoint, test, data })`

This makes it possible to be much more explicit about a `visit` call:

```js
visit(
  (path) => console.log(path),
  {
    entryPoint: 'css',
    test: (path) => path.includes('grid')
  }
)
```